### ISSUE TITLE
Core ML Conversion

### ISSUE THREAD

**Issue by ProfessorHT** [roboflow/rf-detr#191]
Hi, and thank you for your work on this great project! 

I was wondering if there’s any method to convert this to Core ML. Any guidance or pointers would be much appreciated.

Thank you

**Comment by wdewww** [comment]
i tried doing coreml conversion here is my code

```
def export_coreml(output_dir, model, input_shape,inputs_names):

    import coremltools as ct
    export_name = "inference_model"
    if hasattr(model, "export"):
        model.export()

    example_input = torch.randn(*input_shape)  # create an actual tensor
    traced_model = torch.jit.trace(model.eval(), example_input)
    coreml_model = ct.convert(
        traced_model,
        inputs=[ct.TensorType(name=inputs_names[0],shape=input_shape)],
    )

    coreml_model.save(f"{output_dir}/{export_name}.mlpackage")
```
But i got the error Core ML only supports tensors with rank <= 5. Layer "sampling_offsets_1_cast_fp16", with type "reshape", outputs a rank 6 tensor.

**Comment by ProfessorHT** [comment]
Hello,

Any updates regarding this question?

Thank you

**Comment by pzoltowski** [comment]
@nok Is there specific reason why that draft/PR got deleted? I'm getting 404. Previously commented on that PR and it was working well with great speedup. Now I wanted to test with new nano models.

**Comment by menmob** [comment]
> Our supported pathway is via [roboflow-swift](https://github.com/roboflow/roboflow-swift). A blog demonstrating is here: https://blog.roboflow.com/ios-rf-detr-nano/

Am I correct that this repo only downloads a CoreML version of the model from the API and does not have any of the actual conversion code? If so, does this mean the CoreML conversion code will not be published/officially supported?

**Comment by ProfessorHT** [comment]
Hello @timnielen,

Thank you for sharing this solution. I was wondering if there is any library that allows loading RFDETR Core ML model locally in Swift without requiring an API key, since the Roboflow Swift SDK seems to support loading only via an API key.

**Comment by timnielen** [comment]
I‘m not quite sure if I understand your question correctly but the current solution allows you to simply drag and drop the resulting inference_model.mlpackage into xcode. 
Xcode will then generate the appropriate classes automatically. So you can use the model directly.

**Comment by timnielen** [comment]
Hi @ProfessorHT,
I‘m not quite sure if I understand your question correctly but the current solution allows you to simply drag and drop the resulting inference_model.mlpackage into xcode.
Xcode will then generate the appropriate classes automatically. So you can use the model directly.

**Comment by ProfessorHT** [comment]
Hi @timnielen,

Thank you for your quick response!

My goal is to build an iOS application using my own trained model, similar to what’s shown in this [Roboflow tutorial](https://blog.roboflow.com/ios-rf-detr-nano/).

The challenge is that the Roboflow workflow requires me to upload the model online and use their API, but I’d like my app to run completely offline even when there’s no internet connection.

That’s why I’m looking for a way to integrate my model directly, without depending on the Roboflow API.

Thanks again for your guidance!

**Comment by timnielen** [comment]
Hi @ProfessorHT,

Okay yeah in that case it is exactly how I explained. If you use the export_coreml function that I added in my fork (on the develop branch) then you will get an "inference_model.mlpackage" which you can simply drag and drop into xcode. (no upload to roboflow)
Then Xcode will automatically generate the required classes for you to use the model locally. It takes a CGImage as an input and will return two mlshapedarrays. One for the logits and one for the bboxes.

This tutorial by apple might be a good starting point: https://developer.apple.com/documentation/Vision/recognizing-objects-in-live-capture.
Or even use the tutorial from roboflow but simply use your local model instead of the roboflow api.

**Comment by hex41434** [comment]
Hi @timnielen !

which rfdetr version you have installed? I can not convert to coreml. 

rfdetr 1.3.0 does not support preview model.... 

thnx
